src/cephadm/design_choices.rst

```diff
@@ -0,0 +1,165 @@
+
+* disk added/removed
+* daemon added/removed
+* host network/firewall etc has changed
+
+
+Currently, cephadm scrapes each host (up to 10 in parallel[0]) every 6 minutes[1], unless a refresh is forced manually.
+
+Refreshes for disks (ceph-volume), daemons (podman/docker), etc, happen in sequence.
+
+While the refresh model isn't ideal in the first place, the execution time vastly increases with the cluster size > 10
+
+This certainly isn't acceptable since ceph itself scales massively and so has the management part.
+
+There are a couple of solutions to this:
+
+
+1. Pre fetching/caching data on the host
+----------------------------------------
+
+This would allow bin/cephadm to autonomously scrape data that can later be retrieved by mgr/cephadm.
+
+i.e. `cephadm ls` would access an underlying (transparent) cache.
+
+This, btw assumes a **pull** model.
+
+There is at least one possible solutions on *how* to do this.
+
+1.1 daemonize bin/cephadm
+--------------------------
+
+A daemon is able to execute scraping tasks more frequently (as opposed to only when asked by mgr/cephadm)
+
+
+Downsides:
+
+We're still `scp'ing` bin/cephadm from the ceph-mgr to the host which is then executed.
+A daemon on the other hand needs to run continuously on the target system. If ceph-mgr (and the containing bin/cephadm
+script) is updated, the daemon needs to be informed of this event and get updated/restarted as well.
+
+
+1.1.1 HTTP:
+___________
+
+There is a implementation by Paul Cuzner [2] that uses HTTP to serve an endpoint to the hosts metadata.
+
+Upside:
+
+* We could utilize a proper REST API to CRUD all of the daemons/disks etc.
+  Currently it feels a bit weird to call a binary from a remote host (that's another topic though)
+
+
+Downside:
+
+* This is quite some shift in architecture.
+* We still use SSH to bootstrap a node (in order to install bin/cephadm for the first time)
+  Having two distinct transport layers feels odd.
+* The current approach of delivering bin/cephadm to the host doesn't allow the use of external dependencies.
+  This means that we're stuck with the built-in HTTP server lib, which isn't meant for production purposes.
+  bin/cephadm needs to be packaged and distributed (one way or the other) for us to make use of a production ready
+  http server.
+
+Requirements:
+
+* https
+* authentication
+* documentation
+* api versioning (keep it stable)
+
+
+1.1.2 Keep bin/cephadm
+______________________
+
+Daemonizing bin/cephadm allows us to pre-fetch data and store it in some shape or form.
+
+Currently we're only interested in speeding up metadata retrieval. This could mean that we don't even
+need to introduce a new transport layer but rather transparently return cached content via the `cephadm ls`
+or the `cephadm ceph-volume ls` command.
+
+The technical details need to be figured out. (format of data storage etc.)
+
+Upside:
+
+* No new transport layer.
+  Also means no extra auth, security layer is needed
```

Comment:
again - it's unclear as to why 'exporter' would need security but node-exporter and mgr/prometheus do not.

Reply:
They'd still be a different transport layer between the cephadm binary and the daemon though right?

In any case, sticking with ssh also comes with a latency during the ssh connection establishment. I tried to quantify this in this gist - https://gist.github.com/pcuzner/fb2c3c0a9afe70890a55c5244760b511

Looks like a simple cephadm -h takes 1.6s
`[root@rhs-srv-01 ~]# python3 cephadmremote.py                                                                                                                                                                                                 
_load_script starting @ 2020-09-29 18:38:16.479694                                                                                                                                                                                            
_load_script ended @ 2020-09-29 18:38:16.480712                                                                                                                                                                                               
_load_script elapsed time: 0:00:00.001018                                                                                                                                                                                                     
_get_connection starting @ 2020-09-29 18:38:16.480781                                                                                                                                                                                         
_get_connection ended @ 2020-09-29 18:38:17.445635                                                                                                                                                                                            
_get_connection elapsed time: 0:00:00.964854                                                                                                                                                                                                  
exec starting @ 2020-09-29 18:38:17.445812                                                                                                                                                                                                    
INFO:rhs-srv-01.ceph-dev.lab.eng.rdu2.redhat.com:Running command: /usr/bin/python3 -u                                                                                                                                                         
exec ended @ 2020-09-29 18:38:18.166756                                                                                                                                                                                                       
exec elapsed time: 0:00:00.720944 
elapsed time to connect and execute a cephadm -h : 0:00:01.617621`

A curl to the exporter takes 18ms, to return all the metadata
`
# time curl http://rhs-srv-01:5003/v1/metadata -o /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 39921    0 39921    0     0  6497k      0 --:--:-- --:--:-- --:--:-- 6497k

real    0m0.018s
user    0m0.008s
sys     0m0.007s
`

Reply:
> again - it's unclear as to why 'exporter' would need security but node-exporter and mgr/prometheus do not.

Not sure where you read that we don't need https for the prometheus bits too. This is certainly an issue that should be addressed. A major difference is though that cephadm has an encrypted transport already in place.

> In any case, sticking with ssh also comes with a latency during the ssh connection establishment.

Correct, and there are well known approaches to minimize the number times ssh tunnels need to be established.

Reply:
If we're really really concerned about performance and scalability I'll trow [zmq](https://zeromq.org/) in the ring.

Reply:
@jan--f prometheus doesn't currently offer TLS, auth etc does it https://prometheus.io/docs/operating/security/? 

Minimising ssh tunnels sounds good - but you still need an ssh session to every host, for the data gathering right. reducing the tunnel cost would apply to requests once the initial session is open right?

Reply:
The node_exporter got a tls mode in version 1.0.0. https://github.com/prometheus/node_exporter/tree/master/https
Prometheus itself can authenticate to tls endpoints, but still requires a reverse proxy for the server component.

I don't have any measurements for ssh session re-use.